R/plot.R

```diff
@@ -405,100 +368,62 @@ plot_heatmap <- function(scores,
-                             ncol = NULL,
-                             scales = "free_y",
-                             allow_truth_without_pred = FALSE,
-                             remove_from_truth = c("model", "forecaster", "quantile", "prediction", "sample", "interval")) {
+                             range = c(0, 50, 90)) {
 
   # preparations ---------------------------------------------------------------
   # check data argument is provided
-  if (is.null(data) && (is.null(truth_data) | is.null(forecasts))) {
-    stop("need arguments 'data' in function 'score()', or alternatively 'forecasts' and 'truth_data'")
-  }
-
   if (is.null(data)) {
-    data <- merge_pred_and_obs(forecasts, truth_data, by = merge_by, join = "full")
-    if (nrow(data) == 0) {
-      warning("After attempting to merge, only an empty data.table was left")
-      return(data)
-    }
+    stop("need arguments 'data' in function 'score()")
   }
 
   # split truth data and forecasts in order to apply different filtering
   truth_data <- data.table::as.data.table(data)[!is.na(true_value)]
   forecasts <- data.table::as.data.table(data)[!is.na(prediction)]
 
+  del_cols <-
+    colnames(truth_data)[!(colnames(truth_data) %in% c(by, "true_value", x))]
 
-  # function to filter forecast and truth data
-  filter_df <- function(data, filter_list) {
-    data <- data.table::copy(data)
-    # filter as specified by the user
-    for (expr in filter_list) {
-      data <- data[eval(parse(text = expr)), ]
-    }
-    return(data)
-  }
-
-  truth_data <- filter_df(truth_data, c(filter_both, filter_truth))
-  forecasts <- filter_df(forecasts, c(filter_both, filter_forecasts))
-
-  # if specified, get all combinations of the facet variables present in the
-  # forecasts and filter the truth_data accordingly
-  if (!allow_truth_without_pred && !is.null(facet_formula)) {
-    facet_vars <- all.vars(facet_formula)
-    index <- colnames(forecasts)[(colnames(forecasts) %in% facet_vars)]
-    combinations_forecasts <- unique(data.table::copy(forecasts)[, ..index])
-    data.table::setkey(combinations_forecasts)
-    data.table::setkey(truth_data)
-
-    # keep part where predictions are na so they don't get removed by merging
-    truth_without_pred <- truth_data[is.na(prediction)]
-    truth_data <- merge(truth_data, combinations_forecasts)
-    # add back together
-    truth_data <- data.table::rbindlist(list(truth_without_pred, truth_data),
-      use.names = TRUE
-    )
-  }
-
-  # delete certain columns that denominate the forecaster from the truth data
-  truth_data <- delete_columns(truth_data, remove_from_truth, make_unique = TRUE)
+  truth_data <- delete_columns(
+    truth_data,
+    del_cols,
+    make_unique = TRUE
+  )
 
   # find out what type of predictions we have. convert sample based to
   # range data
-  colnames <- colnames(forecasts)
-  if ("sample" %in% colnames) {
+  prediction_type <- get_prediction_type(data)
```

Comment:
How much of this duplicated from score and elsewhere? Can it share an internal function?